src-java/reroute-topology/reroute-storm-topology/src/main/java/org/openkilda/wfm/topology/reroute/service/OperationQueueService.java

```diff
@@ -0,0 +1,135 @@
+
+import org.openkilda.messaging.command.CommandData;
+import org.openkilda.messaging.info.InfoData;
+import org.openkilda.wfm.topology.reroute.bolts.OperationQueueCarrier;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Collection;
+import java.util.Deque;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class OperationQueueService {
+    private final OperationQueueCarrier carrier;
+
+    private final Map<String, FlowQueueData> flowCommands = new HashMap<>();
+
+    public OperationQueueService(OperationQueueCarrier carrier) {
+        this.carrier = carrier;
+    }
+
+    /**
+     * Add command to the front of the queue.
+     */
+    public void addFirst(String flowId, String correlationId, CommandData data) {
+        log.info("Flow command {} has been added to the queue for flow {} with correlation id {}",
+                data, flowId, correlationId);
+        FlowQueueData queueData = getFlowQueueData(flowId);
+        queueData.getQueue().addFirst(new OperationData(correlationId, data));
+        processQueue(queueData);
+    }
+
+    /**
+     * Add command to the end of the queue.
+     */
+    public void addLast(String flowId, String correlationId, CommandData data) {
+        log.info("Flow command {} has been added to the queue for flow {} with correlation id {}",
+                data, flowId, correlationId);
+        FlowQueueData queueData = getFlowQueueData(flowId);
+        queueData.getQueue().addLast(new OperationData(correlationId, data));
+        processQueue(queueData);
+    }
+
+    /**
+     * Handle timeout.
+     */
+    public void handleTimeout(String correlationId) {
+        log.warn("Operation with correlationId {} has timed out", correlationId);
+        Collection<FlowQueueData> foundQueueData = flowCommands.values().stream()
+                .filter(flowQueueData -> correlationId.equals(flowQueueData.getTaskInProgress()))
+                .collect(Collectors.toList());
+
+        if (foundQueueData.isEmpty()) {
+            log.warn("No queue data with correlationId {} found. Timeout event skipped.", correlationId);
+        } else if (foundQueueData.size() > 1) {
+            log.error("Found more than one queue data with correlationId {}. Timed out all of them.", correlationId);
+        }
+
+        foundQueueData.forEach(queueData -> operationCompleted(queueData, null));
+    }
+
+    public void operationCompleted(String flowId, InfoData response) {
+        operationCompleted(getFlowQueueData(flowId), response);
+    }
+
+    private void operationCompleted(FlowQueueData queueData, InfoData response) {
+        log.info("Flow command {} has been completed with response {}", queueData.getTaskInProgress(), response);
```

Comment:
Do we need `success` field in PathSwapResult only to log it here? We already logged it in H&S.